### ISSUE TITLE
Using RITEway in an EcmaScript module environment?

### ISSUE THREAD

**Issue by StaticNoiseLog** [paralleldrive/riteway#322]
I cannot get past errors in my attempts to use RITEway in an EcmaScript modules project (not CommonJS).

This demonstrates the problem:
https://github.com/StaticNoiseLog/riteway-with-esm
The line `"type": "module",` in package.json tells Node that these are EcmaScript modules.
The code in question is `math.test.js`.

First I tried the import syntax (that works for "Tape"):
`import { describe } from 'riteway'`
-> SyntaxError: Named export 'describe' not found. 

Then the workaround suggested by Node:
`import pkg from 'riteway'`
`const { describe } = pkg`
-> TypeError: describe is not a function
I suspected that this might be because I use a style without semicolons. But adding semicolons everywhere did not help.

I am not really familiar with JavaScript and its ecosystem, but as far as I understand it should be possible to use CommonJS modules in EcmaScript modules.
I wonder why I cannot get this to work with RITEway.
Is it possible to use RITEway with EcmaScript modules without Babel or other tooling that I would like to avoid?
Would it make sense to publish RITEway as an EcmaScript module or is that not a good idea?

**Comment by redbar0n** [comment]
I'm also trying to use RITEway outside of a React and Babel context, with as little tooling/compilation as possible (I'm using it to write local tests for a Google Apps Script project). 

I'm getting:

`SyntaxError: Named export 'Try' not found` when I try to `import { Try } from 'riteway'`.

But more importantly, I get:

```
import { describe } from 'riteway'
         ^^^^^^^^
SyntaxError: Named export 'describe' not found. The requested module 'riteway' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'riteway';
const { describe } = pkg;
```

If I try the import suggestion mentioned in the end there, I get the `TypeError: describe is not a function`.

I also tried to make 'riteway' identify as an ES Module, by modifying the `package.json` of `riteway` to include `"type": "module",`. Since RITEway seems to be written using ES Modules internally, and it simply exports to CommonJS by using [esm](https://github.com/standard-things/esm)). But when I do that, then I get:

```
import { describe } from 'riteway'
         ^^^^^^^^
SyntaxError: The requested module 'riteway' does not provide an export named 'describe'
```

**Comment by ericelliott** [comment]
The problem here is that Node.js chose NOT to make ES6 module imports backwards-compatible with common.js the way that literally every compiler in the ecosystem chose to make them work. That decision has made it extremely difficult to upgrade packages built using those compilers to:

1. Continue working with those compilers with no breaking changes, and 
2. Also work with Node's native ES6 module solution.

RITEway currently works fine with every major compiler. Compile your node code, and you can use RITEWay without issue.

But we have not yet found a solution to support Node's native ES6 module solution without breaking existing apps that rely on RITEway in their build step.

We are open to a pull request if you can get it working without breaking changes.

**Comment by ericelliott** [comment]
Maybe we can fix this with `--experimental-specifier-resolution=node`?

**Comment by rauschma** [comment]
If I install OP’s project, I’m seeing in `node_modules/riteway/source/index.js`:

```js
require = require('esm')(module);
module.exports = require('./riteway.js');
```

`esm` simulates ES modules via the CommonJS infrastructure. It looks like it breaks interop with ESM: If I log what’s exported in the second line, I get an empty object.

I’m not surprised – ESM changes a lot of things internally and I’ve heard a few stories of it causing issues.

On the plus side – given that riteway already uses `esm` – it should be easy to switch to real ESM. Or riteway switches to real CommonJS; then you should also be able to import it from ESM.

**Comment by rauschma** [comment]
Proof that the switch to ESM should be relatively easy: I could get the code to work by only changing two things.

First, change the import in `riteway-with-esm/math.test.js`:

```js
import { describe } from 'riteway/source/riteway.js'
```

Second, add this line to `riteway/package.json`:

```json
"type": "module",
```

**Comment by ericelliott** [comment]
@rauschma Thank you very much for taking the time to look into this, but I don't think these suggested fixes will work for us.

## Convert to real CommonJS is a Breaking Change

Named imports from real CommonJS modules are not supported in ESM modules. Instead, you have to import like:

```
import riteway from 'riteway';
const { describe } = riteway;
```

This would not be a terribly big deal except that it will break thousands of build pipelines. 👎 

 ## Convert to ESM
 
Importing EJS in CommonJS is even more problematic. This would break all the CommonJS build pipelines.

Is there any way to export both from one package, or should we deprecate this package and create a new ESM native package?

**Comment by rauschma** [comment]
> Named imports from real CommonJS modules are not supported in ESM modules.

You have to use the right syntax: https://2ality.com/2022/10/commonjs-named-exports.html

> Is there any way to export both from one package

Yes, that’s possible. Do a web search for “hybrid esm commonjs package”.

**Comment by ericelliott** [comment]
> Yes, that’s possible. Do a web search for “hybrid esm commonjs package”. The key is to get the package exports right. I haven’t done this myself, though – all my packages are either CommonJS or ESM, never both.

Yeah, I've done this dance before. I can't find an option that won't break existing build pipelines. It looks like we're stuck deprecating this package and creating a new one.

**Comment by rauschma** [comment]
Maybe:

* Copy `riteway/source/` to `riteway/esm/`
* Add `riteway/esm/package.json` with only one property:
  ```json
  {
    "type": "module"
  }
  ```
* ESM entry point in `"exports"` of `riteway/package.json` is `riteway/esm/riteway.js`

I’m not sure if that works, but it should.

I have also seen people maintain both CommonJS and ESM but the latter was a major version upgrade. Their reasoning: same package, but a breaking change.

**Comment by ericelliott** [comment]
@rauschma Thanks! I think this is the best path forward. Those who want to use it in a native ESM environment can import from `riteway/esm/`. I've released an RC for you guys to test. Please let me know if it works.

This is now available using `npm install --save-dev riteway@latest`.

**Comment by rauschma** [comment]
Wouldn’t it have to be `riteway/esm/index.js` (ESM doesn’t support directory imports)?

You could use package exports to change that to `riteway` or ``riteway/esm`.

**Comment by ericelliott** [comment]
Yeah, I'd like to mirror the import instructions for the existing packages, so all you need to do to use the esm is insert `esm/`. The imports are actually `riteway/esm/riteway.js`, `riteway/esm/match.js`, etc. If I understand it correctly the spec actually requires the `.js` extensions, right?

**Comment by rauschma** [comment]
Ah, OK, makes sense!

**Comment by ericelliott** [comment]
@rauschma Thank you again for your help. It's really appreciated. 🙏

**Comment by ericelliott** [comment]
Should be fixed in riteway@v7.0.0
